core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFileSupplier.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.profiles;
+
+import java.nio.file.Path;
+import java.time.Clock;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.profiles.internal.ProfileFileRefresher;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+/**
+ * Encapsulates the logic for supplying either a single or multiple ProfileFile instances.
+ * <p>
+ * Each call to the {@link #getProfileFile()} method will result in either a new or previously supplied profile based on the
+ * implementation's rules.
+ */
+@SdkPublicApi
+@FunctionalInterface
+public interface ProfileFileSupplier extends SdkAutoCloseable {
+
+    /**
+     * @return A ProfileFile instance.
+     */
+    ProfileFile getProfileFile();
+
+    @Override
+    default void close() {
+    }
+
+    /**
+     * Creates a {@link ProfileFileSupplier} capable of producing multiple profile objects from a file. See
+     * {@link ProfileFileSupplier#builder()} to create a customized implementation.
```

Comment:
Can we add more details about what it does? Something like "Returns a new `ProfileFile` when it is modified on disk"

Reply:
Done.